Support using object of type memoryview, bytearray, struct.Struct, or array.array

ccazabon opened the pull request:
Thanks, this is a very useful package.  I've been using it with buffers/objects of other types, while the existing code only works with ctypes objects.  This change makes `IOC_TYPECHECK` calculate the size correct for memoryview, bytearray, struct.Struct, and array.array in addition to ctypes objects.

vpelletier commented:
Looks good to me.

Superficially, I prefer to avoid `if`/`elif` chains, and `isinstance` is not pythonic, but I see no clean way around here: there is AFAIK no way to get the interpreter to tell us what the size of the buffer would be if we were to extract it (in a `PyArg_Parse` sense), and even less what the size of the buffer of an instance of a type would be (for `ctypes` classes). Creating a memoryview just for this seems wasteful, and seems to come with [garbage-collection considerations](https://docs.python.org/3/library/stdtypes.html#memoryview.release).

There are three docstrings which explicitly mention `(ctype type or instance)`, that I think would benefit from being updated in the same change. Would you mind listing the accepted types in `IOC_TYPECHECK` and referring to it from the three other functions' docstrings ?

ccazabon commented:
I agree that the handling of multiple types is kind of ugly with the conditionals and isinstance checks, but it seemed the least bad option vs. a bunch of chained exception handlers where it just blindly tried `len(t)`, then `t.size`, then ... etc.

I've updated the docstrings and added a bit to the readme.
